lib/client/reexec.go

```diff
@@ -0,0 +1,135 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+// ForkAuthenticateParams are the parameters to RunForkAuthenticate.
+type ForkAuthenticateParams struct {
+	// GetArgs gets the arguments to re-exec with, excluding the executable
+	// (equivalent to os.Args[1:]).
+	GetArgs func(signalFd uint64) []string
+	// Stdin is the child process' stdin.
+	Stdin io.Reader
+	// Stdout is the child process' stdout.
+	Stdout io.Writer
+	// Stderr is the child process' stderr.
+	Stderr io.Writer
+}
+
+type forkAuthCmd struct {
+	*exec.Cmd
+	disownSignal io.ReadCloser
+	parentStdin  io.Reader
+	childStdin   io.WriteCloser
+}
+
+// RunForkAuthenticate re-execs the current executable and waits for any of
+// the following:
+//   - The child process exits (usually in error).
+//   - The child process signals the parent that it is ready to be disowned.
+//   - The context is canceled.
+func RunForkAuthenticate(ctx context.Context, params ForkAuthenticateParams) error {
+	cmd, err := buildForkAuthenticateCommand(params)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return runForkAuthenticateChild(ctx, cmd)
+}
+
+func buildForkAuthenticateCommand(params ForkAuthenticateParams) (*forkAuthCmd, error) {
+	executable, err := os.Executable()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cmd := exec.Command(executable)
+	// Prevent unfinished IO from blocking the child's exit.
+	cmd.WaitDelay = 3 * time.Second
+	// Set up disown signal.
+	signalR, signalW, err := os.Pipe()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	signalFd := addSignalFdToChild(cmd, signalW)
+
+	cmd.Args = append(cmd.Args, params.GetArgs(signalFd)...)
+	cmd.Stdout = params.Stdout
+	cmd.Stderr = params.Stderr
+
+	// Stdin needs to go through an explicit pipe so we can cut it off without
+	// actually closing stdin.
```

Comment:
Stdin most definitely needs to be the exact same stdin so all prompts (such as MFA) can work as intended, and the child can stop using it before requesting to be disowned instead, we should not do an explicit copy (and this would not prevent the child from using the controlling terminal via `/dev/tty`, anyway).